src/main/java/de/tum/in/www1/artemis/service/ExerciseService.java

```diff
@@ -775,4 +786,114 @@ public void validateScoreSettings(Exercise exercise) {
             throw new BadRequestAlertException("The provided bonus points are not allowed", "Exercise", "bonusPointsInvalid");
         }
     }
+
+    /**
+     * Checks the exercise structured grading instructions if any of them is associated with the feedback
+     * then, sets the corresponding exercise field
+     *
+     * @param gradingCriteria grading criteria list of exercise
+     * @param exercise exercise to update
+     *
+     */
+    public void checkExerciseIfStructuredGradingInstructionFeedbackUsed(List<GradingCriterion> gradingCriteria, Exercise exercise) {
+        List<Feedback> feedback = feedbackRepository.findFeedbackByExerciseGradingCriteria(gradingCriteria);
+
+        if (!feedback.isEmpty()) {
+            exercise.setGradingInstructionFeedbackUsed(true);
+        }
+    }
+
+    /**
+     * Re-evaluates the exercise before saving
+     * 1. The feedback associated with the exercise grading instruction needs to be updated
+     * 2. After updating feedback, result needs to be re-calculated
+     *
+     * @param exercise exercise to re-evaluate
+     * @param deleteFeedbackAfterGradingInstructionUpdate  boolean flag that indicates whether the associated feedback should be deleted or not
+     *
+     */
+    public void reEvaluateExercise(Exercise exercise, boolean deleteFeedbackAfterGradingInstructionUpdate) {
+
+        List<GradingCriterion> gradingCriteria = exercise.getGradingCriteria();
+        // retrieve the feedback associated with the structured grading instructions
+        List<Feedback> feedbackToBeUpdated = feedbackRepository.findFeedbackByExerciseGradingCriteria(gradingCriteria);
+
+        // collect all structured grading instructions into the list
+        List<GradingInstruction> gradingInstructions = gradingCriteria.stream().flatMap(gradingCriterion -> gradingCriterion.getStructuredGradingInstructions().stream())
+                .collect(Collectors.toList());
```

Comment:
```suggestion
                .toList();
```
This is better since it only generates a list that can't be changed any more, which is faster